code/modules/deathmatch/deathmatch_lobby.dm

```diff
@@ -0,0 +1,436 @@
+		remove_ckey_from_play(ckey)
+		return
+
+	// equip player
+	var/datum/outfit/deathmatch_loadout/loadout = players[ckey]["loadout"]
+	if (!(loadout in loadouts))
+		loadout = loadouts[1]
+
+	observer.forceMove(loc)
+	var/datum/mind/observer_mind = observer.mind
+	var/mob/living/observer_current = observer.mind?.current
+	var/mob/living/carbon/human/new_player = observer.change_mob_type(/mob/living/carbon/human, delete_old_mob = TRUE)
+	if(!isnull(observer_mind) && observer_current)
+		new_player.AddComponent( \
+			/datum/component/temporary_body, \
+			old_mind = observer_mind, \
+			old_body = observer_current, \
+		)
+	new_player.equipOutfit(loadout) // Loadout
+	players[ckey]["mob"] = new_player
+
+	// register death handling.
+	RegisterSignals(new_player, list(COMSIG_LIVING_DEATH, COMSIG_MOB_GHOSTIZED, COMSIG_QDELETING), PROC_REF(player_died))
+	if (global_chat)
+		ADD_TRAIT(new_player, TRAIT_SIXTHSENSE, INNATE_TRAIT)
+		ADD_TRAIT(new_player, TRAIT_XRAY_HEARING, INNATE_TRAIT)
+
+/datum/deathmatch_lobby/proc/game_took_too_long()
+	if (!location || QDELING(src))
+		return
+	announce(span_reallybig("The players have took too long! Game ending!"))
+	end_game()
+
+/datum/deathmatch_lobby/proc/end_game()
+	if (!location)
+		CRASH("Location of deathmatch game [host] deleted during game.")
+	var/mob/winner
+	if(players.len)
+		var/list/winner_info = players[pick(players)]
+		if(!isnull(winner_info["mob"]))
+			winner = winner_info["mob"] //only one should remain anyway but incase of a draw 
+	
+	announce(span_reallybig("THE GAME HAS ENDED.<BR>THE WINNER IS: [winner ? winner.real_name : "no one"]."))
+	
+	for(var/ckey in players)
+		var/mob/loser = players[ckey]["mob"]
+		UnregisterSignal(loser, list(COMSIG_MOB_GHOSTIZED, COMSIG_QDELETING))
+		players[ckey]["mob"] = null
+		loser.ghostize()
+		qdel(loser)
+
+	game.clear_location(location)
+	game.remove_lobby(host)
+	log_game("Deathmatch game [host] ended.")
+
+/datum/deathmatch_lobby/proc/player_died(mob/living/player, gibbed)
+	SIGNAL_HANDLER
+	if(isnull(player) || QDELING(src))
+		return
+
+	var/ckey = player.ckey
+	if(!islist(players[ckey])) // potentially the player info could hold a reference to this mob so we can figure the ckey out without worrying about ghosting and suicides n such
+		for(var/key in players)
+			var/list/player_info = players[key]
+			if(player_info["mob"] && player_info["mob"] == player)
+				ckey = key
+				break
+	
+	if(!islist(players[ckey])) // if we STILL didnt find a good ckey
+		return
+
+	players -= ckey
+
+	var/mob/dead/observer/ghost = !player.client ? player.get_ghost() : player.ghostize() //this doesnt work on those who used the ghost verb
+	if(!isnull(ghost))
+		add_observer(ghost, (host == ckey))
+	
+	announce(span_reallybig("[player.real_name] HAS DIED.<br>[players.len] REMAIN."))
+
+	if(!gibbed && !QDELING(player) && !istype(player.loc, /obj/effect/abstract)) // for some reason dusting or deleting in chasm storage messes up tgui bad
```

Comment:
Should probably get the other bug fixed rather than putting in half measures in other places in the code